src/transformers/models/UDOP/modeling_udop.py

```diff
@@ -0,0 +1,2860 @@
+        labels: Optional[Tensor] = None,
+        masked_lm_labels: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        char_ids: Optional[Tensor] = None,
+        char_seg_data: Optional[Tensor] = None,
+        inputs_embeds: Optional[Tensor] = None,
+        decoder_inputs_embeds: Optional[Tensor] = None,
+        decoder_head_mask: Optional[Tensor] = None,
+        cross_attn_head_mask: Optional[Tensor] = None,
+        use_cache=True,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        input_dict: Dict[str, Any] = None,
+        **kwargs,
+    ) -> Tuple[Tensor, ...]:
+
+        if seg_data is not None:
+            seg_data = torch.clip(seg_data, 0.0, 1.0)
+        if input_dict is not None:
+            return_task_outputs = []
+            for task in input_dict:
+                return_task_outputs.append(self.forward(**input_dict[task]))
+            return return_task_outputs
+        if encoder_outputs is None:
+            # compute positional bias (can be aggregation of 1D and 2D biases)
+            encoder_position_bias = self.relative_bias(attention_mask=attention_mask, seg_data=seg_data)
+            encoder_outputs = self.encoder(
+                input_ids=input_ids,
+                attention_mask=attention_mask,
+                inputs_embeds=inputs_embeds,
+                head_mask=head_mask,
+                output_attentions=output_attentions,
+                output_hidden_states=output_hidden_states,
+                return_dict=return_dict,
+                position_bias=encoder_position_bias,
+                seg_data=seg_data,
+                image=image,
+                ids_keep=ids_keep,
+            )
+
+        if encoder_outputs is None:
+            return None
+
+        if masked_lm_labels is not None:
+            labels = masked_lm_labels
+
+        if decoder_input_ids is None and labels is not None:
+            decoder_input_ids = self._shift_right(labels)
+
+        if encoder_outputs.vision_embeds is not None:
+            vision_embeds = self.decoder.vision_fc(encoder_outputs.vision_embeds)
+            vision_embeds = self.decoder.vision_norm(vision_embeds)
+            hidden_states = torch.cat([encoder_outputs.last_hidden_state, vision_embeds], 1)
+            encoder_outputs.last_hidden_state = hidden_states
+            encoder_outputs.vision_embeds = None
+
+        # ugly hack for model to work as an encoder
+        if decoder_input_ids is None and labels is None:
+            return encoder_outputs
+
+        attention_mask = torch.cat(
+            [
+                attention_mask,
+                torch.ones_like(
+                    encoder_outputs.last_hidden_state[
+                        :, : encoder_outputs.last_hidden_state.size(1) - attention_mask.size(1), 0
+                    ]
+                ),
+            ],
+            1,
+        )
```

Comment:
It would be helpful to have this be a bit more descriptive (`right_mask` and `right_slice` are probably a terrible names - please pick better ones! :) ) and done in several steps rather than a one liner which is then split across many lines

```suggestion
        right_slice = encoder_outputs.last_hidden_state.size(1) - attention_mask.size(1)
        right_mask = torch.ones_like(encoder_outputs.last_hidden_state[:, :right_slice, 0])
        attention_mask = torch.cat([attention_mask, right_mask], 1)
```